src/serialize.hpp

```diff
@@ -0,0 +1,151 @@
+// limitations under the License.
+
+#ifndef SRC_SERIALIZE_HPP_
+#define SRC_SERIALIZE_HPP_
+
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+template<typename Type> class Serializable{
+public:
+    static void SerializeImpl(const Type& in, std::vector<uint8_t>& out)
+    {
+        size_t nSize = sizeof(in);
+        out.reserve(out.size() + nSize);
+        for (size_t i = 0; i < nSize; ++i) {
+            out.push_back(*((uint8_t*)&in + i));
+        }
+    }
+    static size_t DeserializeImpl(const std::vector<uint8_t>& in, Type& out, size_t position)
+    {
+        size_t size = sizeof(out);
+        if (position + size > in.size()) {
+            throw std::invalid_argument("DeserializeImpl: Trying to read out of bounds.");
+        }
+        for (size_t i = 0; i < size; ++i) {
+            *((uint8_t*)&out + i) = in[position + i];
+        }
+        return size;
+    }
+};
+
+template<typename TypeIn, typename TypeOut>
+void Serialize(const TypeIn& in, TypeOut& out)
+{
+    Serializable<TypeIn>::SerializeImpl(in, out);
+}
+
+template<typename TypeIn, typename TypeOut>
+size_t Deserialize(const TypeIn& in, TypeOut& out, const size_t position)
+{
+    return Serializable<TypeOut>::DeserializeImpl(in, out, position);
+}
+
+template<typename TypeIn, typename TypeOut>
+void SerializeContainer(const TypeIn& in, TypeOut& out)
+{
+    Serialize(in.size(), out);
+    for (auto& entry : in) {
+        Serialize(entry, out);
+    }
+}
+
+template<typename TypeIn, typename TypeOut>
+size_t DeserializeContainer(const TypeIn& in, TypeOut& out, const size_t position)
+{
+    size_t elements;
+    size_t size = Deserialize(in, elements, position);
+    if (elements == 0) {
+        return size;
+    }
+    out.clear();
+    out.resize(elements);
+    for (size_t i = 0; i < elements; ++i) {
+        size += Deserialize(in, out[i], position + size);
+    }
+    return size;
+}
+
+template<typename Type>
+class Serializable<std::vector<Type>>{
+public:
+    static void SerializeImpl(const std::vector<Type>& in, std::vector<uint8_t>& out)
+    {
+        SerializeContainer(in, out);
+    }
+    static size_t DeserializeImpl(const std::vector<uint8_t>& in, std::vector<Type>& out, const size_t position)
+    {
+        return DeserializeContainer(in, out, position);
+    }
+};
+
+template<>
+class Serializable<std::string>{
+public:
+    static void SerializeImpl(const std::string& in, std::vector<uint8_t>& out)
+    {
+        SerializeContainer(in, out);
+    }
+    static size_t DeserializeImpl(const std::vector<uint8_t>& in, std::string& out, const size_t position)
+    {
+        return DeserializeContainer(in, out, position);
+    }
+};
+
+class Serializer
+{
+    std::vector<uint8_t> data;
+public:
+    template <typename InputType>
+    friend Serializer& operator <<(Serializer& serializer, const InputType& value) {
+        Serialize(value, serializer.data);
+        return serializer;
+    }
```

Comment:
I think it would be really nice to constrain this operator as well. If you would add a `static bool specialized = true` to all your `Serializable` classes, you could `std::enable_if<Serializable<InputType>::specialized>` here, to make sure this overload only exists for valid argument types

Reply:
Hm but then we would need to add a specialized version also for all trivial types, no?

Reply:
with mt example above, `is_specialized` would be true for all trivial types (as well as `std::vector` and `std::string`).
So you could constrain it like this:

```
    friend Serializer& operator <<(Serializer& serializer, std::enable_if_t<Serializable<InputType>::specialized, InputType> const& value) {
        Serialize(value, serializer.data);
        return serializer;
    }
```